### PR TITLE
Test against Cucumber ASTs

### PR DESCRIPTION
**Pull request by ciaranmcnulty**
This is a replacement for #152 that checks we are parsing gherkin files the same way that Cucumber does. 

Tests that parse differently are marked as Not Implemented, with an indication of the issue. Hopefully we can work through them one by one.

The tests are pulled from Cucumber's test suite. We can of course add more (and maybe cucumber would appreciate edge cases upstream)

### CODE REVIEW

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php [hunk 1]
```diff
@@ -0,0 +1,194 @@
+<?php
+
+namespace Behat\Gherkin\Loader;
```
Comment (ciaranmcnulty) [inline src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:3]: This is in the main code rather than a test folder because

a. it follows where the Yaml loader is
b. it may be useful if we ever want to outsource the parser to the gherkin binary

**Comment by ciaranmcnulty** [comment]
sigh, this worked fine on my machine :)

File: tests/Behat/Gherkin/Cucumber/CompatibilityTest.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    public function testFeaturesParseTheSameAsCucumber(\SplFileInfo $file)
+    {
+        if (isset((self::NOT_COMPATIBLE)[$file->getFilename()])){
+            $this->markTestIncomplete((self::NOT_COMPATIBLE)[$file->getFilename()]);
+        }
+
+        $gherkinFile = $file->getPathname();
+
+        $actual = $this->normaliseFeature($this->parser->parse(file_get_contents($gherkinFile), $gherkinFile));
+        $expected = $this->normaliseFeature($this->loader->load($gherkinFile . '.ast.ndjson')[0]);
```
Comment (stof) [inline tests/Behat/Gherkin/Cucumber/CompatibilityTest.php:74]: this `[0]` breaks for `empty.feature` which has no feature node.

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+                static function ($child) {
+                    return new BackgroundNode(
+                        $child['background']['name'],
+                        self::getSteps(isset($child['background']['steps']) ? $child['background']['steps'] : []),
+                        $child['background']['keyword'],
+                        $child['background']['location']['line']
+                    );
+                },
+                array_filter(
+                    $json['children'] ? $json['children'] : [],
```
Comment (stof) [inline src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:129]: should be `isset($json['children']) ? $json['children'] : [],` to support cases where the key is not there (without first triggering a notice about non-existent key, which gets turned into an error on CI)

**Review by stof — CHANGES_REQUESTED** [review]
@ciaranmcnulty could it be that your local setup ignores notices about undefined keys ?

File: composer.json [hunk 4]
```diff
@@ -43,5 +44,20 @@
...
         }
-    }
+    },
+
+    "repositories": [
+        {
+            "type": "package",
+            "package": {
+                "name": "cucumber/cucumber",
+                "version": "dev-gherkin-16.0.0",
```
Comment (stof) [inline composer.json:54]: for composer to work properly during a composer update, we need to ensure that update the reference for such a package repository is either updating the version of updating the `time` field for a dev version. Otherwise, composer will consider that your vendor folder is already uptodate (even though it does not have the same ref). Is this something corresponding to some tagged version ?

File: composer.json [hunk 5]
```diff
@@ -43,5 +44,20 @@
...
+                "source": {
+                    "type": "git",
+                    "url": "https://github.com/cucumber/cucumber.git",
+                    "reference": "00f128c8dc2970a564d8a706567836700a7039f0"
```
Comment (stof) [inline composer.json:58]: you might want to add a `dist` config as well, to support downloading the archive rather than cloning (better for caching). See https://github.com/scssphp/scssphp/blob/master/composer.json for an example.

**Comment by stof** [comment]
should we add a separate test dealing with the files from the `bad` folder and expecting a gherkin exception ?

**Comment by ciaranmcnulty** [comment]
@stof thanks for the feedback. Yeah, it seems like the image I was using has a different `error_reporting` - I can work through the rest.

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php (hunk 3 above)
Comment (ciaranmcnulty) [reply src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:129]: I can't wait for `??` to be available to use in this project :)

File: composer.json (hunk 4 above)
Comment (ciaranmcnulty) [reply composer.json:54]: This 'real' tag is `cucumber-gherkin/v16.0.0` but I couldn't seem to get that working

Comment (ciaranmcnulty) [reply composer.json:54]:
```
Could not parse version constraint cucumber-gherkin/v16.0.0: Invalid version string "cucumber-gherkin/v16.0.0"
```

Comment (stof) [reply composer.json:54]: that's indeed not a valid composer version

Comment (ciaranmcnulty) [reply composer.json:54]: Not sure how to handle it when that's the actual tag, any hints? 

https://github.com/cucumber/cucumber/releases/tag/cucumber-gherkin%2Fv16.0.0

Comment (stof) [reply composer.json:54]: Well, as that's an internal thing, we don't care if composer considers that as a dev version rather than a stable version for min stability. And as we reference a tag, we *will* update the `version` field when changing the reference (as we will replace that `16.0.0` part). So the current way is OK.

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+{
+
+    public function supports($resource)
+    {
+        return is_string($resource);
+    }
+
+    public function load($resource)
+    {
+        $json = \json_decode(file_get_contents($resource), true);
```
Comment (stof) [inline src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:28]: this looks wrong to me. NDJson is about having one JSON object per line. you cannot json decode the whole file, otherwise it works only for single-line files.

**Review by stof — CHANGES_REQUESTED** [review]

File: tests/Behat/Gherkin/Cucumber/CompatibilityTest.php [hunk 7]
```diff
@@ -62,10 +62,11 @@ public function setUp()
     /**
      * @dataProvider cucumberFeatures
      */
-    public function testFeaturesParseTheSameAsCucumber(\SplFileInfo $file)
+    public function testFeaturesParseTheSameAsCucumber($file)
```
Comment (stof) [inline tests/Behat/Gherkin/Cucumber/CompatibilityTest.php:65]: why removing this typehint ?

File: tests/Behat/Gherkin/Cucumber/CompatibilityTest.php [hunk 8]
```diff
@@ -82,11 +83,15 @@ public function testFeaturesParseTheSameAsCucumber(\SplFileInfo $file)
 
     public static function cucumberFeatures()
     {
-        foreach (new \FilesystemIterator(self::TESTDATA_PATH . '/good') as $file) {
-           if ($file->isFile() && $file->getExtension() === 'feature') {
-                yield $file->getFilename() => [$file];
+        $files = array();
```
Comment (stof) [inline tests/Behat/Gherkin/Cucumber/CompatibilityTest.php:86]: this is not an array of files but an array of test cases

Comment (stof) [reply tests/Behat/Gherkin/Cucumber/CompatibilityTest.php:86]: also, yield is perfectly fine on PHP 5.6 so you could revert this

Comment (ciaranmcnulty) [reply tests/Behat/Gherkin/Cucumber/CompatibilityTest.php:117]: The lowest version of phpunit we support doesn't allow a generator here

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php (hunk 6 above)
Comment (ciaranmcnulty) [reply src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:28]: That's true; but currently all the files are single-line! Maybe it's enough to add an assertion that there's only one line?

Comment (stof) [reply src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:28]: but this file is named NDJsonLoader, not JsonLoader

File: tests/Behat/Gherkin/Cucumber/CompatibilityTest.php (hunk 8 above)
Comment (stof) [reply tests/Behat/Gherkin/Cucumber/CompatibilityTest.php:117]: that's only a bug in 5.7.0, which was fixed in 5.7.1. Bump the min version to 5.7.1 for the 5.7 branch (or better to the latest 5.7.x patch release) and you're OK.

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php (hunk 6 above)
Comment (stof) [reply src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:28]: btw, supporting multiple lines would not be an issue: the method already reports an array, so parsing each (non-empty) line with `json_decode` and `self::getFeature` would do the work.

Comment (ciaranmcnulty) [reply src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:28]: That's probably the best approach, thanks

I guess on the test side I can cast from` FeatureNode[]` to `?FeatureNode` (erroring if there's >1) ?

Comment (stof) [reply src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:28]: Well, for now, yes, because our parser does not support multiple features per parsed file (btw, the description in the phpdoc says it returns a feature array, which is weird in this regard)

Comment (ciaranmcnulty) [reply src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:28]: That's because it's implemented ultimately by things like DirectoryLoader

Comment (ciaranmcnulty) [reply src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php]: OK now using `file()`

**Comment by ciaranmcnulty** [comment]
I think I addressed the feedback now

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+{
+
+    public function supports($resource)
+    {
+        return is_string($resource);
+    }
+
+    public function load($resource)
+    {
+        return array_map(
```
Comment (stof) [inline src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:26]: you should either filter out `null` values or forbid `getFeature` from returning it

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php [hunk 10]
```diff
@@ -0,0 +1,193 @@
...
+{
+
+    public function supports($resource)
+    {
+        return is_string($resource);
+    }
+
+    public function load($resource)
+    {
+        return array_values(array_map(
```
Comment (stof) [inline src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:26]: isn't this missing a `array_filter` ?

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+                    }
+                )
+            )
+        );
+    }
+
+    /**
+     * @return BackgroundNode|null
+     */
+    private static function getBackground($json)
```
Comment (stof) [inline src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:115]: missing `array` typehint

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php (hunk 10 above)
Comment (ciaranmcnulty) [reply src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:26]: Yep

**Comment by ciaranmcnulty** [comment]
@stof changes addressed I think

File: src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php [hunk 12]
```diff
@@ -23,7 +23,7 @@ public function supports($resource)
 
     public function load($resource)
     {
-        return array_values(array_map(
+        return array_filter(array_map(
```
Comment (stof) [inline src/Behat/Gherkin/Loader/CucumberNDJsonAstLoader.php:26]: I would actually still keep the `array_values` call so that we return a list rather than a array with holes.

**Comment by stof** [comment]
@ciaranmcnulty I suggest creating issues for the various parity issues (if they don't exist yet)

**Comment by ciaranmcnulty** [comment]
@stof yeah after it's merged I'll raise issues (so I can hyperlink to the exclusions in the issue descriptions)

File: tests/Behat/Gherkin/Cucumber/CompatibilityTest.php [hunk 13]
```diff
@@ -0,0 +1,165 @@
...
+     * @var Parser
+     */
+    private $parser;
+
+    /**
+     * @var LoaderInterface
+     */
+    private $loader;
+
+    public function setUp()
```
Comment (stof) [inline tests/Behat/Gherkin/Cucumber/CompatibilityTest.php:61]: should be `protected` to match the visibility in PHPUnit

**Review by stof — COMMENTED** [review]
I suggest cleaning the history of this PR before the merge by squashing commits which are about small review fixes (or partial reverts of changes done in a previous commit of the PR)

**Comment by ciaranmcnulty** [comment]
@stof default in this repo is to squash and merge at the end, so I wasn't keeping the history too clean. I'm happy to tidy up if you think

**Comment by stof** [comment]
well, if you do a squash at the merging time, then this already counts as keeping the history clean (I don't care about the history inside the PR view, but in the merged history). I was not aware of the fact you always squash at merging time.

**Comment by ciaranmcnulty** [comment]
Only because someone set the repo to that being the default, at some point :)

**Comment by ciaranmcnulty** [comment]
Thanks for all the review as always, @stof. It is much appreciated!
